main.go

```diff
@@ -130,31 +148,63 @@ func HandleRun(w http.ResponseWriter, req *http.Request) {
 	err = shell.StdIO(nil, pts, pipe)
 	if err != nil {
 		log.Println(err)
-		return
+		return output, err
 	}
-	err = shell.Run(runCtx, req.Body)
+	err = shell.Run(runCtx, req)
 	if err != nil {
 		log.Println(err)
 	}
 
-	b, _ := json.Marshal(struct {
-		Dir            string
-		Stdout, Stderr string
-		Err            error
-	}{
-		shell.Dir(),
-		string(terminal.Render(stdout.Bytes())),
-		stderr.String(),
-		err,
-	})
-
-	_, err = w.Write(b)
-	if err != nil {
-		log.Println(err)
-	}
+	output.Dir = shell.Dir()
+	output.Stdout = string(terminal.Render(stdout.Bytes()))
+	output.Stderr = stderr.String()
+	output.Err = err
+	return output, nil
+}
+
+func HandleRun(w http.ResponseWriter, req *http.Request) {
+	go func() {
+		output, err := Run(req.Body)
+		if err != nil {
+			log.Println(err)
+			return
+		}
+		log.Println(req.Body)
+		o, err := json.Marshal(output)
+		if err != nil {
+			log.Println(err)
+		}
+		_, err = w.Write(o)
+		if err != nil {
+			log.Println(err)
+		}
+	}()
 }
 
 func HandleCancel(w http.ResponseWriter, req *http.Request) {
 	log.Print("Received cancel")
 	runCancel()
 }
+
+// currently bash complete only, wrapper around "Run"
+func Complete(req io.Reader) (string, error) {
+	buf, err := ioutil.ReadAll(req)
+	if err != nil {
+		log.Println(err)
+		return "", err
+	}
+	prog, _, hasArgs := strings.Cut(string(buf[:]), " ")
+	if hasArgs {
+		comm := fmt.Sprintf("complete -p '%s' | sed s/^complete/compgen/", prog)
+		log.Println(comm)
+		completion, _ := Run(strings.NewReader(comm))
```

Comment:
You shouldn't need to piggyback off of Shell.Run() like this. Also, bash-specific code goes in bash.go.

A better way might be to:
1. Define an abstract Completions() method in the Shell interface, which takes a partial command and returns a list of completions. Use that here, instead of calling Run.
2. Add a new method to command_server.sh which runs `complete` and prints the results in a way that can be easily parsed.
3. In bash.go, implement Completions() by calling the new method and parsing results.

Reply:
Thanks, yeah that sounds much cleaner!